### PR TITLE
feat(posthog-recordings): address three errors found in replay capture

### PR DESCRIPTION
**Pull request by pauldambra**
if a replay event is too large for our kafka (and that means it is _really_ large) then we drop it
currently we log that to loki so we can see it is happening
but we don't tell the SDK or the user

this changes the capture pathway for replay only so that it replaces the message with a new one instead of dropping it

* adding a custom event to the recording so we know we dropped data
* reject payloads that cause a `ValueError` when they're being processed (instead of swallowing the error)
* when there is a kafka timeout error we return that to the client as a 504 (the web sdk retries those)

dropping payloads with either full or incremental snapshots will mean playback might be at best odd and at worst impossible

* added developer tests
* checked event ingestion and replay ingestion still working locally

### CODE REVIEW

File: posthog/api/capture.py [hunk 1]
```diff
@@ -534,14 +541,25 @@ def get_event(request):
...
                 start_time = time.monotonic()
                 for future in futures:
-                    future.get(timeout=settings.KAFKA_PRODUCE_ACK_TIMEOUT_SECONDS - (time.monotonic() - start_time))
-
+                    if future is not None:
+                        future.get(timeout=settings.KAFKA_PRODUCE_ACK_TIMEOUT_SECONDS - (time.monotonic() - start_time))
+    except ValueError as e:
+        with sentry_sdk.push_scope() as scope:
+            scope.set_tag("capture-pathway", "replay")
+            scope.set_tag("token", token)
```
Comment (marandaneto) [inline posthog/api/capture.py:555]: what is the token here? is it something we'd know while debugging and searching at sentry?

Comment (pauldambra) [reply posthog/api/capture.py:555]: yep, or rather we'd know it debugging in posthog
it lets us see what team was affected

Comment (pauldambra) [reply posthog/api/capture.py]: will update tag name for the future traveller

File: posthog/api/capture.py [hunk 2]
```diff
@@ -551,6 +569,89 @@ def get_event(request):
...
+                            "tag": "Message too large",
+                        },
+                        "$window_id": first_item.get("$window_id"),
+                        "timestamp": first_item.get("timestamp"),
+                    }
+                ],
+            },
+        }
+    except Exception as ex:
+        capture_exception(ex)
```
Comment (marandaneto) [inline posthog/api/capture.py:609]: I'd love that all events captured during replay ingestion would have a tag such as:
scope.set_tag("capture-pathway", "replay")
so we could easily query/have a dashboard for all errors happening in this part.

**Review by daibhin — APPROVED** [review]

**Review by marandaneto — APPROVED** [review]

File: posthog/api/capture.py [hunk 3]
```diff
@@ -534,23 +541,122 @@ def get_event(request):
...
                 start_time = time.monotonic()
                 for future in futures:
-                    future.get(timeout=settings.KAFKA_PRODUCE_ACK_TIMEOUT_SECONDS - (time.monotonic() - start_time))
-
+                    if future is not None:
+                        future.get(timeout=settings.KAFKA_PRODUCE_ACK_TIMEOUT_SECONDS - (time.monotonic() - start_time))
+    except ValueError as e:
+        with sentry_sdk.push_scope() as scope:
+            scope.set_tag("capture-pathway", "replay")
+            scope.set_tag("ph-team-token", token)
```
Comment (marandaneto) [inline posthog/api/capture.py:555]: I am almost sure this is gonna be removed by Sentry automatically because of data scrubbing regexes (...token).

Comment (pauldambra) [reply posthog/api/capture.py:555]: 🤞

**Comment by pauldambra** [comment]
@marandaneto do mobile SDKs retry on 504?

**Comment by marandaneto** [comment]
> @marandaneto do mobile SDKs retry on 504?

The answer is it depends, on if the HTTP client (okthtp) treats it as a [client error](https://github.com/PostHog/posthog-android/blob/b4747f9854c1e7e0c444f848ecd738984c9cac00/posthog/src/main/java/com/posthog/internal/PostHogUtils.kt#L25), it does.
If the server returns 504 before the client times out, no [sadly](https://github.com/PostHog/posthog-android/blob/b4747f9854c1e7e0c444f848ecd738984c9cac00/posthog/src/main/java/com/posthog/internal/PostHogQueue.kt#L324-L341).
Usually retrying every time leads to an infinite loop and fills up the device's cache which is even worse so that's the best we can do, of course we can retry if it's 504 specifically, we do that with 413 already.
408, 503,  would be probably another ones, etc.

File: posthog/api/capture.py [hunk 4]
```diff
@@ -534,23 +541,135 @@ def get_event(request):
...
+            capture_exception(e)
+        # this means we're getting an event we can't process, we shouldn't swallow this
+        # in production this is mostly seen as events with a missing distinct_id
+        return cors_response(
+            request,
+            generate_exception_response("capture", f"Invalid recording payload", code="invalid_payload"),
+        )
+    except KafkaTimeoutError as kte:
```
Comment (bretthoerner) [inline posthog/api/capture.py:563]: Mostly out of curiosity, are you actually observing these (often?) or this just being defensive?

**Review by bretthoerner — APPROVED** [review]
Makes sense to me.

File: posthog/api/capture.py (hunk 4 above)
Comment (pauldambra) [reply posthog/api/capture.py:563]: as many as 20 a minute (if I'm reading [this log output correctly](https://grafana.prod-us.posthog.dev/explore?schemaVersion=1&panes=%7B%22d2u%22:%7B%22datasource%22:%22P8E80F9AEF21F6940%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bcontainer%3D%5C%22posthog-recordings%5C%22%7D%20%7C%3D%20%60KafkaTimeoutError%60%20%7C%3D%20%60kafka_session_recording_produce_failure%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22P8E80F9AEF21F6940%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%22now-3h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1))

my guess is these are messages that are just way too big for us to process and we're timing out before we're discovering how big they are but 🤷

Comment (pauldambra) [reply posthog/api/capture.py:577]: which got me wondering about what we'd do if the data for some session can only time out

i've added a probe for the web sdk's retry_count parameter to try and avoid trying infinitely and DDOSing ourselves 

feels a bit yucky but might protect us

(@marandaneto might be useful if iOS and Android added `retry_count` as a query parameter when sending replays but no rush 😊)

**Pull request merged**